staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go

```diff
@@ -166,3 +166,56 @@ type TracingConfiguration struct {
 	// Defaults to 0.
 	SamplingRatePerMillion *int32
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowControlConfig holds flow control parameters.
+type FlowControlConfig struct {
+	metav1.TypeMeta
+
+	// WorkEstimatorConfig holds work estimator parameters.
+	// +optional
+	*WorkEstimatorConfig
+}
+
+// WorkEstimatorConfig holds work estimator parameters.
+type WorkEstimatorConfig struct {
+	// ListWorkEstimatorConfig holds work estimator parameters related to list requests.
+	// +optional
+	ListWorkEstimatorConfig *ListWorkEstimatorConfig
+	// MutatingWorkEstimatorConfig holds work estimator
+	// parameters related to watches of mutating objects.
+	// +optional
+	MutatingWorkEstimatorConfig *MutatingWorkEstimatorConfig
+
+	// MaximumSeats is the maximum number of seats a request can occupy
+	//
+	// NOTE: work_estimate_seats_samples metric uses the value of maximumSeats
+	// as the upper bound, so when we change maximumSeats we should also
+	// update the buckets of the metric.
```

Comment:
I do not see a metric by that name.  Have I overlooked something, or is https://github.com/kubernetes/apiserver/blob/fc846fd4280ddfedde77e11f4f79e3be356401d5/pkg/util/flowcontrol/metrics/metrics.go#L294-L306 the intended referent?

As far as I know, the buckets of a histogram can not be dynamically adjusted.  The way we have handled a dynamic upper bound in other situations is with a combination of (a) a gauge of the upper bound and (b) a histogram of observations of the ratio of (the more dynamic interesting variable, here estimated seats) / (upper bound).

Reply:
This note was copied from it's original position at `width.go` - [PR #106628](https://github.com/kubernetes/kubernetes/pull/106628/files#diff-5e38ba69dd8edd7eb8570342352c203bb5f3115a9650496697de3dfb7f210307).

Reply:
I think this comment applies to `work_estimated_seats` metric and this is indeed problematic as it has `10` harcoded. This probably should be made configurable too.

Reply:
> This probably should be made configurable too.

Are you suggesting something different from what was suggested in my remark above?

Reply:
No, you are definitely right about having this gauge metric per priority level. I just got stuck with implementing this change, the code is non-trivial for me.

If I understand correctly, `NewTimingRatioHistogramVec` needs to be initialised first for each priority level and only then values may be added to it. But the code adding `workEstimate.MaxSeats()` lies in `pkg/server/filters` where it doesn't have access to either `flowcontrol.priorityLevelState` or `apiserver.FlowControlConfig` instances. I am not sure how to proceed with this.

Reply:
I guess the only way to do this is to move the metric handling from `pkg/server/filters/priority-and-fairness.go` ([src](https://github.com/kubernetes/kubernetes/blob/fdd921cad0cd9308ec62c1b86c9c1cc5d12e5d21/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L135)) to `flowcontrol.configController.startRequest` ([src](https://github.com/kubernetes/kubernetes/blob/fdd921cad0cd9308ec62c1b86c9c1cc5d12e5d21/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go#L817))?

Reply:
I decided to stick with `NewHistogramVec` for now. Since this is a change I'm unsure about, I added it in a separate commit within this PR (to be squashed later): https://github.com/kubernetes/kubernetes/pull/111222/commits/8c47afb94b921e3a1cce305b9efd5bac07631631. Please let me know what do you think about it. :)